micropip/package_index.py

```diff
@@ -0,0 +1,121 @@
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any
+
+from packaging.utils import (
+    parse_wheel_filename,
+)
+from packaging.version import InvalidVersion, Version
+
+
+@dataclass
+class ProjectInfoFile:
+    filename: str  # Name of the file
+    url: str  # URL to download the file
+    version: Version  # Version of the package
+    sha256: str  # SHA256 hash of the file
+
+    # Size of the file in bytes, if available (PEP 700)
+    # This key is not available in the Simple API HTML response, so this field may be None
+    size: int | None = None
+
+
+@dataclass
+class ProjectInfo:
+    """
+    This class stores common metadata that can be obtained from different APIs (JSON, Simple)
+    provided by PyPI. Responses received from PyPI or other package indexes that support the
+    same APIs must be converted to this class before being processed by micropip.
+    """
+
+    name: str  # Name of the package
+
+    # List of releases available for the package, sorted in ascending order by version
+    # This list only contains wheels not sdist, but it does not filter by tags for it might contain
+    # binary wheels for other platforms. So it is up to the caller to filter the list by tags.
+    # TODO: should we filter by tags here? Or should we leave it to the caller?
+    releases: dict[Version, list[ProjectInfoFile]]
+
+    @staticmethod
+    def from_json_api(data: dict[str, Any]) -> "ProjectInfo":
+        """
+        Parse JSON API response
+
+        https://warehouse.pypa.io/api-reference/json.html
+        """
+
+        name: str = data.get("info", {}).get("name", "UNKNOWN")
+        _releases: dict[str, Any] = data["releases"]
+
+        releases: dict[Version, list[ProjectInfoFile]] = defaultdict(list)
+        for version_str, fileinfo in _releases.items():
+            try:
+                version = Version(version_str)
+            except InvalidVersion:
+                # Ignore non PEP 440 compliant versions
+                continue
+
+            for file in fileinfo:
+                filename = file["filename"]
+                if not filename.endswith(".whl"):
```

Comment:
Would it make sense to filter the compatible python versions / platform here? I mean keeping all those wheels that cannot be used is not very useful.

Reply:
Yes, I mentioned that in the comment above. One concern is that usually, we don't need to iterate through the whole list of versions because, when we do `micropip.install` it will often only look at the latest version.

I'll check if the filtering process is fast enough. If not I'll try using a generator.